### PR TITLE
[Offload] Only initialize a plugin if it is needed

### PR DESCRIPTION
**Pull request by jhuber6**
Summary:
Initializing the plugins requires initializing the runtime like CUDA or
HSA. This has a considerable overhead on most platforms, so we should
only actually initialize a plugin if it is needed by any image that is
loaded.

### CODE REVIEW

File: offload/plugins-nextgen/common/src/PluginInterface.cpp [hunk 1]
```diff
@@ -1578,14 +1580,27 @@ Expected<bool> GenericPluginTy::checkELFImage(StringRef Image) const {
...
-  // Perform plugin-dependent checks for the specific architecture if needed.
-  return isELFCompatible(Image);
+  LLVMContext Context;
+  auto ModuleOrErr = getLazyBitcodeModule(MemoryBufferRef(Image, ""), Context,
+                                          /*ShouldLazyLoadMetadata=*/true);
+  if (!ModuleOrErr)
+    return ModuleOrErr.takeError();
+  Module &M = **ModuleOrErr;
+
+  return Triple(M.getTargetTriple()).getArch() == getTripleArch();
```
Comment (jplehr) [inline offload/plugins-nextgen/common/src/PluginInterface.cpp:1597]: The change from `getArch()` to `getTripleArch` intended?

Comment (jhuber6) [reply offload/plugins-nextgen/common/src/PluginInterface.cpp:1597]: Yes

File: offload/src/PluginManager.cpp [hunk 2]
```diff
@@ -160,6 +153,27 @@ void PluginManager::registerLib(__tgt_bin_desc *Desc) {
     if (Entry.flags == OMP_REGISTER_REQUIRES)
       PM->addRequirements(Entry.data);
 
+  // Initialize all the plugins that have associated images.
+  for (auto &Plugin : Plugins) {
+    if (Plugin->is_active())
```
[the remark this answers is not in the document]
Comment (jhuber6) [reply offload/src/PluginManager.cpp:158]: Whether or not the `init` method was called. I couldn't think of a good name for this, it's just a way to control whether or not we can use the plugin to do the actual checking. So, first pass just says "Is this an ELF with a matching platform" second pass says "Does this subarchitecture match".

[the remark this answers is not in the document]
Comment (jplehr) [reply offload/src/PluginManager.cpp:158]: If it indicates that the `init` method was called, why not call `is_initialized`?

**Comment by jplehr** [comment]
Should we close https://github.com/llvm/llvm-project/pull/76174 in favor of this one?

File: offload/src/PluginManager.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (jplehr) [reply offload/src/PluginManager.cpp:158]: Does this now need to be `is_initialized`?

**Review by jplehr — APPROVED** [review]
LGTM

**Comment by jhuber6** [comment]
@ronlieb AMD version https://github.com/llvm/llvm-project/issues/93183. It only required moving some code because AOMP has the logging stuff.

**Pull request merged**